File: building-h2o.cs
using System.Threading;

public class H2O {

    private  Barrier water;
    private Semaphore oxygen;
    private Semaphore hydrogen;
    public H2O() {
        this.water = new Barrier(3);
        this.oxygen = new Semaphore(1, 1);
        this.hydrogen = new Semaphore(2, 2);
    }

    public void Hydrogen(Action releaseHydrogen) {
        hydrogen.WaitOne();
        water.SignalAndWait();
        // releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen();
        hydrogen.Release();
    }

    public void Oxygen(Action releaseOxygen) {
        oxygen.WaitOne();
        water.SignalAndWait();
        // releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen();
        oxygen.Release();
    }
}

File: Concurrency/Java/building-h2o.java
class H2O {

    private Semaphore h2;
    private Semaphore o;
    private CyclicBarrier barrier;
    
    public H2O() {
        h2 = new Semaphore(2);
        o = new Semaphore(1);
        barrier = new CyclicBarrier(3);
    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
        h2.acquire();
        try{
            barrier.await();    
        }catch(BrokenBarrierException ex)
        {
            System.err.println(ex);
        }
        releaseHydrogen.run();
        h2.release();
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        o.acquire();
        try{
            barrier.await();    
        }catch(BrokenBarrierException ex)
        {
            System.err.println(ex);
        }
        releaseOxygen.run();
        o.release();
    }
}
